test/test_pipeline_schedule.py

```diff
@@ -0,0 +1,238 @@
+
+e.g. (2x2, 2 hosts with 2 processes)
+torchrun --rdzv-backend=c10d --rdzv-endpoint=node1.example.com:29400 --nnodes=2 --nproc-per-node=2 test_pipeline_schedule.py
+"""
+
+import argparse
+import logging
+import os
+
+from collections import deque
+from datetime import timedelta
+from typing import List, Optional, Union
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import torch.nn as nn
+from pippy.PipelineSchedule import (
+    PipelineScheduleGPipe,
+    PipelineScheduleLoopedBFS,
+    PipelineScheduleLoopedDFS,
+    PipelineStage,
+)
+from torch.profiler import profile, ProfilerActivity, record_function
+
+logger = logging.getLogger(__name__)
+
+
+class MLP(nn.Module):
+    def __init__(
+        self,
+        dim: int,
+        hidden_dim: int,
+        out_dim: int,
+    ):
+        super().__init__()
+        self.wi = nn.Linear(dim, hidden_dim, bias=False)
+        self.wh1 = nn.Linear(hidden_dim, hidden_dim, bias=False)
+        self.wh2 = nn.Linear(hidden_dim, hidden_dim, bias=False)
+        self.wh3 = nn.Linear(hidden_dim, hidden_dim, bias=False)
+        self.wo = nn.Linear(hidden_dim, out_dim, bias=False)
+        self.gelu_act = nn.GELU(approximate="tanh")
+
+    def forward(self, x):
+        a = self.wi(x)
+        a = self.wh1(a)
+        a = self.wh2(a)
+        a = self.wh3(a)
+        b = self.gelu_act(a)
+        c = self.wo(b)
+        return c
+
+
+def setup(local_rank, world_size):
+    # If this is a child process (i.e., its PID is not the same as the PID of the process that started this script)
+    if os.getppid() != os.getpid():
+        set_up_logging(local_rank)
+
+    # initialize the process group
+    logger.info(f"init for rank {local_rank}")
+    dist.init_process_group("nccl")
+    if torch.distributed.is_initialized():
+        torch.cuda.set_device(local_rank)
+
+    logger.info(f"finish init for rank {local_rank}")
+
+
+def main(**kwargs):
+    torch.manual_seed(42)
+    print(f"MY KWARGS ARE {kwargs}")
+    rank = kwargs["rank"]
+    local_rank = kwargs["local_rank"]
+    world_size = kwargs["world_size"]
+
+    setup(local_rank, world_size)
+    logger.info(
+        f"====== World Rank {rank}, Local Rank {local_rank}, World Size {world_size} main ======"
+    )
+
+    input_dim = 4000
+    hidden_dim = 8000
+    output_dim = 4000
+
+    module_list = torch.nn.ModuleList(
+        modules=[
+            MLP(input_dim, hidden_dim, output_dim) for i in range(world_size)
+        ]
+    )
+    microbatch_size = 8
+    global_batch_size = 64
+    assert global_batch_size % microbatch_size == 0
+    n_microbatches = int(global_batch_size / microbatch_size)
+    n_pp = world_size
+
+    x = torch.randn([microbatch_size, input_dim]).to("meta")
+
+    stage_model = PipelineStage(
+        module_list[rank], rank, world_size, rank, world_size, x
```

Comment:
QQ: why are `rank, world_size` provided twice?

Reply:
The respective arguments are:

        rank -> stage_id: int,
        world_size -> num_stages: int,
        rank -> rank: int,
        world_size -> world_size: int,
        
For the nonlooped schedules the stage_id and num_stages are always equal to rank and world_size